lib/Pulse.Lib.AVLTree.fst

```diff
@@ -0,0 +1,948 @@
+      let node = !np;
+     
+      rewrite each _node as node;
+
+      let left_new = append_left node.left v;
+      
+      np := {node with left = left_new};
+      
+      rewrite each left_new as ({ node with left = left_new }).left in (is_tree left_new ((T.append_left (reveal _ltree) v)));
+      
+      rewrite each node.right as ({ node with left = left_new }).right in (is_tree node.right _rtree);
+       
+      intro_is_tree_node x np;
+
+      x
+    }
+  }
+} 
+```
+
+
+```pulse
+fn rec append_right (#t:Type0) (x:tree_t t) (v:t) (#ft:G.erased (T.tree t))
+requires is_tree x ft
+returns y:tree_t t
+ensures is_tree y  (T.append_right ft v)
+{
+   match x {
+    None -> {
+      
+      is_tree_case_none x;
+      
+      elim_is_tree_leaf x;
+     
+      let left = create t;
+      let right = create t;
+      
+    
+      let y = node_cons v left right;
+      
+      let np = Some?.v y;
+      
+      is_tree_case_some y np;
+
+      intro_is_tree_node y np;
+      
+      y 
+    }
+    Some vl -> {
+      
+      let np = Some?.v x;
+      
+      is_tree_case_some x np;
+      
+      with _node _ltree _rtree._;
+      
+      let node = !np;
+      
+      rewrite each _node as node;
+
+      let right_new = append_right node.right v;
+      
+      np := {node with right = right_new};
+      
+      rewrite each right_new as ({ node with right = right_new }).right in (is_tree right_new ((T.append_right (reveal _rtree) v)));
+      
+      rewrite each node.left as ({node with right = right_new}).left in (is_tree node.left _ltree);
+      
+      intro_is_tree_node x np;
+      
+      x
+    }
+  }
+} 
+```
+
+
+```pulse
+fn node_data (#t:Type) (x:tree_t t) (#ft:G.erased (T.tree t))
+    requires is_tree x ft  ** (pure (Some? x))
+    returns v:t
+    ensures is_tree x ft
+{
+  let np = Some?.v x;
+      
+  is_tree_case_some x np;
+      
+  with _node _ltree _rtree._;
+      
+  let node = !np;
+      
+  rewrite each _node as node;
+
+  let v = node.data;
+  
+  intro_is_tree_node x np;
+  v
+}
+```
+
+```pulse
+fn rec is_mem (#t:eqtype) (x:tree_t t) (v: t) (#ft:G.erased (T.tree t))
+    requires is_tree x ft
+    returns b:bool
+    ensures is_tree x ft ** pure (b <==> (T.mem ft v))
+{
+   match x {
+        None -> {
+            is_tree_case_none x;
+            false
+        }
+        Some vl -> {
+            //let dat = node_data x;
+            is_tree_case_some x vl;
+            with _node _ltree _rtree. _;
+            let n = !vl;
+            rewrite each _node as n;
+
+            let dat = n.data;
+            
+            if (dat = v) 
+            {
+              intro_is_tree_node x vl;
+              true
+            }
+            else{
+              let b1 = is_mem n.left v;
+              let b2 = is_mem n.right v;
+
+              let b3 = b1 || b2;
+              intro_is_tree_node x vl;
+              b3;
+              
+            }
+        }
+    }
+}
+```
+
+module G = FStar.Ghost
```

Comment:
There are multiple of this abbreviation in this file.

Reply:
comment addressed.